stream_settings: Restructure stream creation panel.

sujalshah-bit opened the pull request:
## User group section 

| panel 1 | panel 2 |
| -------- | -------- |
| ![image](https://github.com/zulip/zulip/assets/73663475/cb6b0513-3951-4c7d-8ae2-62d28f1b0cf0) | ![image](https://github.com/zulip/zulip/assets/73663475/1abbadf1-1bf1-426a-8c54-7e6461a033ef)|


-----

## Channel creation section
| panel 1 | panel 2 |
| -------- | -------- |
| ![image](https://github.com/zulip/zulip/assets/73663475/a41b64ab-6c8d-4d6b-b8ee-64bfe595a557) | ![image](https://github.com/zulip/zulip/assets/73663475/d092addf-6a6e-4765-a915-34800f422641) |

- ###  Advance configuration:
| img | img |
| -------- | -------- |
| ![image](https://github.com/zulip/zulip/assets/73663475/8fc8a08d-3c40-471d-a559-7bb196a3a7c3) | ![image](https://github.com/zulip/zulip/assets/73663475/edf94696-9fc4-4e40-ba86-aaa5f3447907) |

-----
### Notes

- The issue description doesn't include a mention of a "Configure Settings" button. While implementing these changes, I realized that users might need to revisit the configuration settings to modify certain details, such as the stream name. Therefore, I added a button for navigating back to the configuration settings when needed.
 
- Also the color of the button should be different I guess please provide feedback on both of this.

-----

Fixes part of #29403.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
Is this ready for review?

sujalshah-bit commented:
@timabbott as i have mentioned above last part that is **Change how adding users works** this is part is still remaining for  that part i still not don't know how to implement this

timabbott commented:
Let's just move the "Change how adding users work" part to a separate issue; this already has enough going on as it is. @sujalshah-bit can you do a careful pass to make sure you've got the right strings? "Advance configuration" in your screenshot looks like a typo, "Advanced" is the word you want.

sujalshah-bit commented:
I will make the update soon.

timabbott commented:
@sahil839 can you review this one?

sujalshah-bit commented:
@timabbott this is ready for review :)

sahil839 commented on web/templates/stream_settings/stream_creation_form.hbs:
```diff
@@ -40,8 +40,15 @@
+            <div class="settings-sticky-footer-right">
+                <button class="create_stream_cancel button small white rounded" data-dismiss="modal">{{t "Cancel" }}</button>
+                <button class="finalize_create_stream button small sea-green rounded hide" type="submit">{{t "Create" }}</button>
+                <button id="create_stream_next_tab" class="button small sea-green rounded">{{t "Continue to add subscribers" }}</button>
```
I think we should use better IDs here instead of using `next_tab` and `back_tab`  and have more specific names like `stream_creation_go_to_subscribers`, etc.

sahil839 commented on web/templates/stream_settings/stream_creation_form.hbs:
```diff
@@ -5,32 +5,32 @@
                     </div>
                 </section>
-                <section class="block">
+                <section class="block stream_create_add_subscriber_container" style="display: none;">
```
It is better to avoid adding inline styles and instead add it in respective CSS files.

sahil839 commented on web/templates/stream_settings/stream_creation_form.hbs:
```diff
@@ -5,32 +5,32 @@
-                <section class="block">
-                    <label for="create_stream_description" class="settings-field-label">
+                <section  class="block create_stream_description_container">
+                    <label for="create_stream_description">
```
Why have you removed this `settings-field-label` class.

sahil839 commented on web/src/stream_create.js:
```diff
@@ -139,6 +139,44 @@ function toggle_advanced_configurations() {
+        // Hide add subscriber container and show stream creation containers
+        $(
+            ".create_stream_title_container, .create_stream_description_container, .create_stream_types_container",
+        ).show();
```
We can probably add a containing element and just hide and show one element.

sahil839 replied:
Also, probably one change can be we can use `show_subs_pane.create_stream`. Since we are also changing the heading we can consider it as we are changing the complete right panel and that can be handled there. We can probably pass a variable to `show_subs_pane.create_stream` with the name of the container to show and probably a separate function to handle the buttons in footer.

sahil839 commented on web/src/user_group_create.js:
```diff
@@ -25,6 +25,46 @@ export function get_name() {
+        $("#groups_overlay #create_user_group_next_tab").show();
+        $("#groups_overlay #create_user_group_back_tab").hide();
+        $("#groups_overlay .user-group-info-title").text(
+            $t({defaultMessage: "Create user group: configure settings"}),
```
These changes can be consistent with what we do for streams.

sujalshah-bit replied:
to be consistent with what we do in streams. I have imported `user_group_edit.js` in `user_group_create.js` but it is giving me error.

```console
eslint    | /home/sujal/zulip/web/src/user_group_create.js
eslint    |   12:1  error  Dependency cycle detected  import/no-cycle
eslint    |
eslint    | /home/sujal/zulip/web/src/user_group_edit.js
eslint    |   28:1  error  Dependency cycle detected  import/no-cycle

```

sahil839 replied:
I think you can move `show_user_group_settings_pane` to `user_group_components` in a prep commit and that should solve this.

sahil839 commented:
Also, the current changes also results in the "Advanced configuration" expandable view in the stream edit form as same template is used for both the places. The issue does not mention anything about changing it. @alya do we want that change in stream edit UI as well?

alya commented:
The section label should be "Advanced configurations", as noted in the issue.

alya commented:
Please be sure not to introduce a regression by using "stream" instead of "channel".

![Screenshot 2024-05-14 at 16 44 54](https://github.com/zulip/zulip/assets/2090066/0a22fd36-2886-403a-bfb9-506670d54f42)

alya commented:
Currently, I can click "Continue to add subscribers" without filling out required info on the first page (stream name). We should make sure to catch that before the user gets to the next page.

alya commented:
> The issue description doesn't include a mention of a "Configure Settings" button. While implementing these changes, I realized that users might need to revisit the configuration settings to modify certain details, such as the stream name. Therefore, I added a button for navigating back to the configuration settings when needed.

Good idea! We might want to make it more clear that this is a "back" button, but the current state on this PR is probably OK for now.

alya commented:
> Also, the current changes also results in the "Advanced configuration" expandable view in the stream edit form as same template is used for both the places. The issue does not mention anything about changing it. @alya do we want that change in stream edit UI as well?

Good question! I wasn't thinking about that, but it feels reasonable to me.

alya commented:
Clicking anywhere on the "Advanced configuration" section name should expand/collapse it, with an appropriate pointer cursor.

alya commented:
I'm seeing the triangle not rotating down correctly when the section is expanded in the "create channel" UI:

![Screenshot 2024-05-14 at 16 58 24](https://github.com/zulip/zulip/assets/2090066/fe7458f7-2d23-45f9-91c6-fc61f4cc9b37)

alya commented:
That's all I've got for now. I didn't read @sahil839 's feedback carefully, so there may be some duplication.

sujalshah-bit commented:
>Currently, I can click "Continue to add subscribers" without filling out required info on the first page (channel name). We should make sure to catch that before the user gets to the next page.
Similarly, it shouldn't be possible to continue to the next page if you use a channel name that's already taken, or there's some other error.

-  I am not able to reproduce that bug. In my dev env it is working as expected
![image](https://github.com/zulip/zulip/assets/73663475/23d19855-5d27-44a4-b0e5-a8db7b205dd7)
![image](https://github.com/zulip/zulip/assets/73663475/c7624875-9f7e-430d-b688-2e9838d51510)


---

>I'm seeing the triangle not rotating down correctly when the section is expanded in the "create channel" UI:

-  This bug only occurs when the user clicks on a stream and then clicks the create stream button. I'm trying to understand what's causing it, but I haven't figured it out much. @sahil839, can you help?



https://github.com/zulip/zulip/assets/73663475/17f9c79a-8a9a-4e85-873a-c253f66e4e88


---

- left one comment above for `user_group` discussion @sahil839 kindly review it.

---

Other than that I think I have covers all feedbacks kindly review @sahil839 @alya.

sahil839 commented:
> I am not able to reproduce that bug. In my dev env it is working as expected

I can reproduce this. I just opened the stream creation form, did nothing and simply clicked on "Continue to add subscribers" button.

> This bug only occurs when the user clicks on a stream and then clicks the create stream button. I'm trying to understand what's causing it, but I haven't figured it out much. @sahil839, can you help?

This is because there are two elements with same ID "toggle-advanced-configurations-icon" once the user clicks on a stream - one for stream creation form and one for the stream settings edit form. The correct solution here would be to use a class for that element and then update the JS code to make sure the class is changed for the correct element.

sahil839 commented on web/src/stream_create.js:
```diff
@@ -139,6 +139,30 @@ function toggle_advanced_configurations() {
+    e.stopPropagation();
+
+    const stream_name = $("#create_stream_name").val().trim();
+    const name_ok = stream_name_error.validate_for_submit(stream_name);
```
A better variable name could be `is_stream_name_valid`.

sahil839 commented on web/src/stream_create.js:
```diff
@@ -139,6 +139,30 @@ function toggle_advanced_configurations() {
+    if (name_ok) {
+        stream_settings_components.show_subs_pane.create_stream("subscribers_container");
+    } else {
+        stream_settings_components.show_subs_pane.create_stream("configure_channel_settings");
```
Can we just do nothing in case the name is not valid?

sahil839 commented on web/src/stream_settings_components.js:
```diff
@@ -45,13 +45,38 @@ export const show_subs_pane = {
     },
 };
 
+export function update_footer_buttons(container_name = "configure_channel_settings") {
```
I think there is no need for a default value as the caller will pass container name always.

sahil839 commented on web/src/user_group_components.ts:
```diff
@@ -81,12 +81,35 @@ export const show_user_group_settings_pane = {
     },
 };
+
+export function update_footer_buttons(container_name = "configure_user_group_settings"): void {
```
Here also, we can remove the default value.

sahil839 commented on web/src/user_group_create.js:
```diff
@@ -89,6 +89,32 @@ class UserGroupNameError {
+    e.stopPropagation();
+
+    const group_name = $("#create_user_group_name").val().trim();
+    const name_ok = user_group_name_error.validate_for_submit(group_name);
```
This variable can be named similar to how we would have it for streams.

sahil839 reviewed (commented):
Posted some comments.

sahil839 commented:
Since we now collapse settings like "Who can post to stream", "Message retention", etc. in the stream edit view as well, do we want to show the "Save changes" button on "Advanced configurations" heading?

![Peek 2024-05-23 12-57](https://github.com/zulip/zulip/assets/35494118/5eea4057-2761-4f0e-bf48-e232749fec1c)

alya commented:
> Since we now collapse settings like "Who can post to stream", "Message retention", etc. in the stream edit view as well, do we want to show the "Save changes" button on "Advanced configurations" heading? @alya FYI.


Sure, yeah, I think that would be an improvement.

alya commented:
@sujalshah-bit Could you please prioritize addressing the feedback here, as it's a priority for the 9.0 release?

sujalshah-bit commented:
I notice some issues regarding the repositioning of "save and discard" buttons to the advanced config heading:

- In stream edit if you change advanced config settings, it will show the buttons. However, when you change one of the channel permission settings (radio buttons), it will not show the "save and discard" buttons.
![save_buttons](https://github.com/zulip/zulip/assets/73663475/960d46ba-7627-478b-bf43-1a92737c0ddf)







- Also, when creating a stream, it is giving an error due to this change.
![save_buttons_errors](https://github.com/zulip/zulip/assets/73663475/5dfffee7-00a8-4e82-83c7-d1ee525ca695)

This is the code I changed:
```hbs
<div class="advanced-configurations-container settings-subsection-parent">{{!-- added settings-subsection-parent here --}}
    <div class="advance-config-title-container subsection-header"> {{!-- added  subsection-header here --}}
        <i class="fa fa-sm fa-caret-right toggle-advanced-configurations-icon" aria-hidden="true"></i>
        <h4 class="stream_setting_subsection_title"><span>{{t 'Advanced configurations' }}</span></h4>
        {{> ../settings/settings_save_discard_widget section_name="stream-permissions" }} {{!-- added  widget here --}}

    </div>
```

sujalshah-bit commented:
> Could you please prioritize addressing the feedback here, as it's a priority for the 9.0 release?

I forgot to mentioned, I have address all the point of @sahil839  except the to show buttons in advance config heading.

sahil839 commented:
I think you might need to have a look at JS code in `settings_components.ts` to how the widget works and you would be able to find why it is not working. 

Regarding the second bug I guess you would just need to make sure that the save-discard widget code is not called for stream creation page.

sahil839 reviewed (commented):
Posted one comment about save discard widget not working correctly.

Did a quick read of code and it seems fine, will do a final review once the save-discard widget behavior is fixed.

sujalshah-bit commented:
### stream privacy
![one](https://github.com/zulip/zulip/assets/73663475/7da81c45-a2f4-4ecd-8609-382c620159b2)


### advance config:
![two](https://github.com/zulip/zulip/assets/73663475/50107573-d0ea-4f78-92ea-b4195b325f04)


@sahil839 updated the PR.

sahil839 commented on web/templates/stream_settings/stream_settings.hbs:
```diff
@@ -53,15 +53,7 @@
-
-                <div class="stream-permissions-warning-banner"></div>
-
+            <div class="stream-permissions" id="stream_permission_settings">
```
Can we skip this adding this `div`? There is already a containing `div` in `stream_types`.

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -1,76 +1,107 @@
-    </label>
+<div class="{{#if is_stream_edit}}settings-subsection-parent{{/if}}">
+    {{#if is_stream_edit}}
+        <div class="{{#if is_stream_edit}}subsection-header{{/if}}">
```
This inner `is_stream_edit` conditions are not needed here I guess.

sahil839 commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -1,76 +1,107 @@
-  widget_name=can_remove_subscribers_setting_widget_name
-  label=(t 'Who can unsubscribe others from this channel?')
-  value_type="number"}}
+<div class="advanced-configurations-container {{#if is_stream_edit}}settings-subsection-parent settings-sujal{{/if}}">
```
Can we just keep these classes hare and below irrespective of whether it is stream creation or stream edit form? Also, there is a typo here I guess with `settings-sujal` class.

sujalshah-bit replied:
I forgot to remove `settings-sujal` 😅sorry for that. If we remove `{{#if is_stream_edit}}` then at the time of stream creation it will give this error :

>[Also, when creating a stream, it is giving an error due to this change.](https://github.com/zulip/zulip/pull/29433#:~:text=and%20discard%22%20buttons.-,Also%2C%20when%20creating%20a%20stream%2C%20it%20is%20giving%20an%20error%20due%20to%20this%20change.,-This%20is%20the)

sahil839 replied:
I think this is because the save-discard widget handler is called for stream creation page as well. Maybe we can update that to only be called for the stream edit page in a prep commit.

We can still keep this condition since it this class is only used for the save-discard handler and nothing else like CSS or something, but the class name is such that it does not seem only related to the save-discard widget someone reading the code might feel why is this class included only for stream settings.

sahil839 replied:
Or anyways might be good to just have separate files for the permissions section and advanced configuration section with heading and save discard widget code being in the `stream_settings.hbs. But I guess that would be good as a follow-up and not worth of blocking this further.

So, I think you can just keep this as it is for now and we can handle it later.

sahil839 commented on web/src/stream_create.js:
```diff
@@ -430,6 +452,7 @@ export function set_up_handlers() {
     $container.on("click", ".finalize_create_stream", (e) => {
         e.preventDefault();
         clear_error_display();
+        stream_settings_components.show_subs_pane.create_stream("configure_channel_settings");
```
Why is this added? Due to this, if I select no subscribers in the "Subscribers" page and click on "Create" button, an error is shown there for not adding subscribers but error is not visible since the view is changed to settings panel.

sujalshah-bit replied:
I added this code initially to handle a situation where the 'name' field might be empty. This would redirect the user to the configuration part. However, I've made some changes to the code, which now resolves this issue. Similarly, I've also addressed the same problem in the 'usergroup' section

sahil839 reviewed (commented):
Posted a few comments.

sujalshah-bit commented:
@sahil839 made the changes and left comment above kindly check :)

sahil839 commented on web/src/stream_create.js:
```diff
@@ -435,16 +457,19 @@ export function set_up_handlers() {
         const principals = stream_create_subscribers.get_principals();
         if (principals.length === 0) {
             stream_subscription_error.report_no_subs_to_stream();
+            stream_settings_components.show_subs_pane.create_stream("subscribers_container");
```
Is this needed, since user will already be on subscribers panel.

The code for `name_ok` part will still be needed to handle the case where some other user creates a stream with same name in the time this user is adding subscribers.

sahil839 reviewed (commented):
Posted one comment.

Also, would be good to check the case how an error returned by server is shown in the UI.

sujalshah-bit commented:
**Insufficient permissions**
![image](https://github.com/zulip/zulip/assets/73663475/d9d8e955-e653-4e7c-80f4-2252b6e5f583)
----
**Unable to access channel "unauthorized_streams_name"**
**You can only invite other Zephyr mirroring users to private channels.**

I DON'T KNOW HOW TO REPLICATE THE ABOVE TWO ERROR.

sahil839 commented:
It is fine for checking one case, just wanted to make sure that the message is shown as expected. Though I am not sure we should show it in the "general settings" section and not in the subscribers section.

sahil839 commented:
The code changes look good mostly.

One thing that I noticed while testing is that the banner does not appear when I try changing privacy of an unsubscribed stream to private.

sujalshah-bit commented:
@sahil839 fix the bug. It was due to `stream_permission_settings` id was not added in the new containing `div`

sahil839 commented:
Looks good to me. I haven't tested the banner part again assuming you would have tested it. @alya do you want to have a look at this again.

I mentioned about one potential code improvement follow-up [here](https://github.com/zulip/zulip/pull/29433#discussion_r1637470105).

sujalshah-bit commented:
Yes this is the ss for banner part:
![image](https://github.com/zulip/zulip/assets/73663475/5849772e-875b-4ce6-a8b7-203b77e7457e)

alya commented:
Thanks! A few more thoughts from playing with the PR:

1. I'm not liking the new channel panel titles. Let's try:
- Configure new channel settings
- Add subscribers to <channel name, with its privacy icon>

2. Let's rename "Configure settings" to "Back to settings".
3. Am I imagining it, or is there too much vertical space here compared to other settings?
![Screenshot 2024-06-14 at 15 55 17](https://github.com/zulip/zulip/assets/2090066/a4a559b6-18b6-4e82-ae77-83fe410fb99c)

alya commented:
Otherwise looks good to me. I'm feeling like we may want to turn the "Who can access the channel?" setting into a dropdown to match everything else better, maybe with explanations like in the Inbox dropdown. But we can think about that as a follow-up.

![Screenshot 2024-06-14 at 16 06 48](https://github.com/zulip/zulip/assets/2090066/f1b58be1-cf84-46e6-9278-7a56078c50d3)

sujalshah-bit commented:
>Am I imagining it, or is there too much vertical space here compared to other settings?

@alya, I've updated the margins for both the "announce stream" and "default stream" to `20px` from `25px`. The changes are reflected in the following images:
|  Before  |  After |
|---|---|
| ![image](https://github.com/zulip/zulip/assets/73663475/5816c597-35fb-487a-9f9a-509e3ba097c2)  |  ![image](https://github.com/zulip/zulip/assets/73663475/a1db8cb1-e38b-4a15-a247-681e526ee467) | 

The changes were introduced by commits [10e6fd0](https://github.com/zulip/zulip/commit/10e6fd04e728f38bdc3a9f7ebef49090d6159c48) and [a817157](https://github.com/zulip/zulip/commit/a81715786c5fdfb031d9372d822ecfa633051064#diff-647172a09930fbcb22d76e929d552227717cad35bc9482465e19d8eb47c7ba2e).

After reviewing this, what are your thoughts on whether we should maintain the current margins or update them?

sujalshah-bit commented:
>I'm not liking the new channel panel titles. Let's try:

|   |   | 
|---|---|
| ![image](https://github.com/zulip/zulip/assets/73663475/69edaf1b-f4f8-4fdd-ae9d-af3d073f8c71)| ![image](https://github.com/zulip/zulip/assets/73663475/5c94aa4a-638c-460e-a33f-d89b12d25fda)  |

made the changes. Also usergroup has similar titles should I also change them to `Configure new group settings` and  `Add members to <Group name>` ?

alya commented:
> Also usergroup has similar titles should I also change them to Configure new group settings and Add members to <Group name> ?

Sounds good!

alya commented:
> After reviewing this, what are your thoughts on whether we should maintain the current margins or update them?

The updated version looks better to me.

alya commented:
@timabbott There's a small open item to rename the group panels, but I think you can take a look at this PR now. Note that this is a blocker for  #29825, so would be great to get it integrated.

timabbott commented:
This looks like a great checkpoint, merged, thanks @sujalshah-bit and @sahil839 for all the work on this!

I have a handful of changes to request in a follow-up:
- We seem to be using `<section id="user-group-permission-container" class="block">`, and using `class=` on the section elements in stream settings; we should probably decide which is better and consolidate on that.
- It seems that these modules use the problematic pattern of declaring click handlers at top-level, rather than in an `initialize` function; this is problematic for testing, among other things, because code gets executed on just importing the modules. We can fix this, probably for the whole collection of violations, in a follow-up PR.

```
$ git grep '^\$\("body'
docs/contributing/code-style.md:$("body").on("click", ".move_message_button", function (e) {
web/src/alert_popup.ts:$("body").on("click", ".alert-box > div .exit", function () {
web/src/stream_create.js:$("body").on("click", ".settings-sticky-footer #stream_creation_go_to_subscribers", (e) => {
web/src/stream_create.js:$("body").on(
web/src/stream_create.js:$("body").on("click", ".advanced-configurations-container .advance-config-title-container", (e) => {
web/src/user_group_create.js:$("body").on("click", ".settings-sticky-footer #user_group_go_to_members", (e) => {
web/src/user_group_create.js:$("body").on("click", ".settings-sticky-footer #user_group_go_to_configure_settings", (e) => {
```
